File: tree/Traversal/Preorder/iterative/python/PreorderTraversal.py
from collections.abc import Sequence


def PreorderTraversal(left_children: Sequence[int], right_children: Sequence[int], root: int) -> list[int]:
    order: list[int] = []
    stack = [root]

    while stack:
        node = stack.pop()
        order.append(node)

        right_child = right_children[node]
        if right_child != -1:
            stack.append(right_child)

        left_child = left_children[node]
        if left_child != -1:
            stack.append(left_child)

    return order

File: tree/Traversal/Preorder/iterative/java/PreorderTraversal.java
import java.util.*;

public class PreorderTraversal {
    public static List<Integer> PreorderTraversal(int[] leftChildren, int[] rightChildren, int root) {
        List<Integer> order = new ArrayList<>();
        Deque<Integer> stack = new ArrayDeque<>();
        stack.push(root);

        while (!stack.isEmpty()) {
            int node = stack.pop();
            order.add(node);

            int rightChild = rightChildren[node];
            if (rightChild != -1) {
                stack.push(rightChild);
            }

            int leftChild = leftChildren[node];
            if (leftChild != -1) {
                stack.push(leftChild);
            }
        }

        return order;
    }
}
